service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/flight/RetrieveAzureResourcesRegionStep.java

```diff
@@ -0,0 +1,242 @@
+      ControlledResource resource,
+      @Nullable String region) {
+    if (region != null) {
+      UUID resourceId = resource.getResourceId();
+      resourceIdToRegionMap.put(resourceId, region);
+      resourceIdToWorkspaceIdMap.put(resourceId, resource.getWorkspaceId().toString());
+    }
+  }
+
+  private String getAzureDiskRegion(
+      ControlledAzureDiskResource resource, AzureCloudContext azureCloudContext) {
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    return computeManager
+        .disks()
+        .getByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getDiskName())
+        .regionName();
+  }
+
+  private String getAzureStorageAccountRegion(
+      ControlledAzureStorageResource resource, AzureCloudContext azureCloudContext) {
+    StorageManager storageManager = crlService.getStorageManager(azureCloudContext, azureConfig);
+
+    return storageManager
+        .storageAccounts()
+        .getByResourceGroup(
+            azureCloudContext.getAzureResourceGroupId(), resource.getStorageAccountName())
+        .regionName();
+  }
+
+  private String getAzureRelayNameSpaceRegion(
+      ControlledAzureRelayNamespaceResource resource, AzureCloudContext azureCloudContext) {
+    RelayManager manager = crlService.getRelayManager(azureCloudContext, azureConfig);
+    return manager
+        .namespaces()
+        .getByResourceGroup(
+            azureCloudContext.getAzureResourceGroupId(), resource.getNamespaceName())
+        .regionName();
+  }
+
+  private String getAzureStorageContainerRegion(
+      ControlledAzureStorageContainerResource resource, AzureCloudContext azureCloudContext) {
+    StorageManager storageManager = crlService.getStorageManager(azureCloudContext, azureConfig);
+    if (resource.getStorageAccountId() != null) {
+      WsmResource wsmResource =
+          resourceDao.getResource(resource.getWorkspaceId(), resource.getStorageAccountId());
+      ControlledAzureStorageResource storageAccount =
+          wsmResource
+              .castToControlledResource()
+              .castByEnum(WsmResourceType.CONTROLLED_AZURE_STORAGE_ACCOUNT);
+      return getAzureStorageAccountRegion(storageAccount, azureCloudContext);
+    }
+    UUID landingZoneId =
+        landingZoneApiDispatch.getLandingZoneId(
+            new BearerToken(userRequest.getRequiredToken()), resource.getWorkspaceId());
+    Optional<ApiAzureLandingZoneDeployedResource> existingSharedStorageAccount =
+        landingZoneApiDispatch.getSharedStorageAccount(
+            new BearerToken(userRequest.getRequiredToken()), landingZoneId);
+    return existingSharedStorageAccount
+        .map(
+            apiAzureLandingZoneDeployedResource ->
+                storageManager
+                    .storageAccounts()
+                    .getById(apiAzureLandingZoneDeployedResource.getResourceId())
+                    .regionName())
+        .orElse(null);
```

Comment:
Should this throw an exception? If we have a container with no storage account ID and no existing shared storage account it seems like we're in uncharted territory and should bail out, or at least log a warning.

Reply:
i don't want to abandon the whole operation because one workspace is "corrupted". logging warning sg.